### PR TITLE
[TF:TRT] Fix profile handling for pruned inputs

### PR DESCRIPTION
**Pull request by tfeher**
Before TRT 8, an input tensor can be pruned if it is not used by the network . This can happen if only the shape of the input tensor is used and the shape of the input tensor is known, that is, min equals max in the shape optimization profile (nvbugs/3153064). This PR fixes optimization profile handling if there are pruned inputs.

Notes:
- The number TRT network input for INetworkDefinition is always the same as the TF network inputs.
- The number of input bindings for a single profile of an ICudaEngine is not larger than the network inputs.

To fix profile handling:
- We add `is_input_pruned_[]` vector to TrtShapeOptimizationProfile. 
- We ensure that in `OptimizationProfileConfig` the `min.size()` is always determined by the number of the network inputs, and not by the number of input bindings.

Tracker: #45481

Tagging @bixia1 for review.

### CODE REVIEW

File: tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc [hunk 1]
```diff
@@ -509,25 +518,35 @@ Status TrtShapeOptimizationProfile::RestoreProfiles(
...
 #endif
   int n_inputs = GetNumberOfEngineInputs(engine);
+  if (n_inputs > n_network_inputs) {
+    return errors::Internal("Incorrect number of engine inputs");
+  }
   VLOG(2) << "Attempting to restore " << n_profiles << " profiles, each with "
           << n_inputs << " inputs";
-  SetShapeTensorMask(engine, n_inputs);
+  SetShapeTensorMask(engine, n_network_inputs);
+  is_pruned_input_.resize(n_network_inputs);
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc:527]: Let's add a comment to the definition of profiles_, to document that if an input, say i,  is pruned, and the profiles is restored from an build, then profiles_[][i] is not initialized.

**Comment by bixia1** [comment]
I have also edit the PR description, please check.

**Review by tfeher — COMMENTED** [review]
Thanks @bixia for the review, I have addressed the issues. We will have to wait merging this until we fix shape output memcpy bug (improved solution in place of IntsOnDevice attr, that we discussed offline) #52276. I will update here once we are ready to proceed.

**Comment by tfeher** [comment]
@bixia1 I have rebased the PR, please have a look.

**Review by bixia1 — APPROVED** [review]

**Review by bixia1 — APPROVED** [review]

**Pull request merged**